docs/proposals/component-registration.md

```diff
@@ -0,0 +1,320 @@
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/component-registration.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Proposal: Component Registration
+
+## Table of Contents
+
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [Proposal: Component Registration](#proposal-component-registration)
+  - [Table of Contents](#table-of-contents)
+  - [Terminology](#terminology)
+  - [Abstract](#abstract)
+  - [Problems](#problems)
+    - [ComponentStatuses](#componentstatuses)
+    - [Endpoints](#endpoints)
+  - [Use Cases](#use-cases)
+    - [Deployment](#deployment)
+    - [Debugging](#debugging)
+    - [Notification](#notification)
+    - [Reporting](#reporting)
+    - [Self-Healing](#self-healing)
+  - [Design](#design)
+    - [API Changes](#api-changes)
+      - [API Types](#api-types)
+    - [Kubectl (CLI)](#kubectl-cli)
+    - [Storage](#storage)
+    - [Status Updates](#status-updates)
+  - [Example](#example)
+  - [Alternate Naming](#alternate-naming)
+  - [Next Steps](#next-steps)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+## Terminology
+
+A Kubernetes Cluster **Component**...
+
+- Is considered a part of the cluster
+- Must register (or be registered) with Kubernetes (apiserver) to join the cluster
+- Must know about Kubernetes (speak the API)
+- Must be deployed/managed by the cluster admin/operator (rather than a user)
+- Should expose a method for remotely probing the liveness and readiness of each instance (e.g. tcp/http/https).
+- May be scheduled/deployed on the cluster (primary components vs addon components)
+- May be required for the cluster to function (required vs optional)
+- May be added/removed from a running cluster (mutable cluster component membership over time)
+- May require admin API access (privileged behavior)
+- May consist of one or more instances
+- May be deployed on different (network accessible) machines
+- May be upgraded over time
+
+The **Primary Components** currently include apiserver, controller-manager, and scheduler. The lifecycle of these components is managed by some external system. These components are required for Kubernetes to function.
+
+The **Addon Components** currently include kube-dns and kube-ui. The lifecycle of these components is managed by the kubernetes cluster that they provide functionality to. These components add functionality, but are not required for Kubernetes to function. Kubernetes currently ships with some of these required for "conformance", but they're technically not required for minimal orchestration and scheduling.
```

Comment:
Heapster would go in the list, yes. The list is going to get stale over time tho. It's less important than the distinction. 

But you're right, if this proposal is implemented then there doesn't need to be an official distinction. There's no API distinction. They're just pods that add functionality to k8s, may or may not have admin access, are probably in the kube-system namespace, and hopefully have a label that identifies them as a "component". 

With real ACLs, a non-admin user probably wouldn't be able to deploy one. But there might be some future where a namespace can have namespaced ACLs to allow a namespace admin to deploy a component that adds functionality constrained to that namespace.